src/native/s3_client.c

```diff
@@ -0,0 +1,389 @@
+        (void)body_response_result;
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onResponseBody callback",
+                (void *)meta_request);
+        }
+    }
+
+    (*env)->DeleteLocalRef(env, jni_payload);
+}
+
+/* only used in S3 client for now, but has generic JNI/HTTP utility */
+static jobjectArray s_aws_java_http_headers_from_native(JNIEnv *env, const struct aws_http_headers *headers) {
+    const size_t header_count = aws_http_headers_count(headers);
+    jobjectArray ret = (jobjectArray)(*env)->NewObjectArray(
+        env, (jsize)header_count, http_header_properties.http_header_class, (void *)NULL);
+
+    for (size_t index = 0; index < header_count; index += 1) {
+        struct aws_http_header header;
+        aws_http_headers_get_index(headers, index, &header);
+        jbyteArray header_name = aws_jni_byte_array_from_cursor(env, &header.name);
+        jbyteArray header_value = aws_jni_byte_array_from_cursor(env, &header.value);
+
+        jobject java_http_header = (*env)->NewObject(
+            env,
+            http_header_properties.http_header_class,
+            http_header_properties.constructor_method_id,
+            header_name,
+            header_value);
+
+        (*env)->SetObjectArrayElement(env, ret, (jsize)index, java_http_header);
+    }
+
+    return (ret);
+}
+
+static void s_on_s3_meta_request_headers_callback(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_http_headers *headers,
+    int response_status,
+    void *user_data) {
+    (void)meta_request;
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        (struct s3_client_make_meta_request_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    if (callback_data->java_s3_meta_request_response_handler_native_adapter != NULL) {
+        jobjectArray java_headers_array = s_aws_java_http_headers_from_native(env, headers);
+
+        /* size_t header_length = aws_array_list_length(&headers->array_list);
+        printf("Headers length: %zu", header_length); */
+
+        AWS_LOGF_TRACE(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p: invoking S3MetaRequest.onResponseHeaders callback. Code: %d",
+            (void *)meta_request,
+            response_status);
+
+        (*env)->CallVoidMethod(
+            env,
+            callback_data->java_s3_meta_request_response_handler_native_adapter,
+            s3_meta_request_response_handler_native_adapter_properties.onResponseHeaders,
+            response_status,
+            java_headers_array);
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onResponseHeaders callback",
+                (void *)meta_request);
+        }
```

Comment:
Need a fuller understanding of JNI clean up. The array likely does, but the unresolved question for me is do all of the array elements inside need separate clean ups as well?

Reply:
Please don’t construct headers from jni and send them into java. I spent a whole week purging the codebase of that this summer. It’s slow at runtime, and error prone send them as a byte[] all the code is already written for you, you just need to use it.

Reply:
The S3 callback in native gives me `struct aws_http_headers`. As long as that contained array_lists's data pointer is the blob I need to pass I'm good, otherwise a bit more work is needed to reconstruct the headers blob, or re-wire the callback from aws-c-s3

Reply:
I don't have access to `headers->array_list.data` if the data layout is usable for this scenario. Only the forward declaration of `struct aws_http_headers` is visible. If that's intentional, or the layout of the array_list.ptr isn't correct then I need to adjust the aws-c-s3 callback to give headers as the entire blob instead.